New Task Timeline Display Setting

hassoncs opened the pull request:
Adds a TaskTimeline into the TaskHeader similar to Cline.

![image](https://github.com/user-attachments/assets/967d492b-88e1-490a-b68c-b3f94fb6b8aa)
![2025-06-18 17 45 49](https://github.com/user-attachments/assets/fe9f1e5e-f2de-49af-a3b8-994f80f06c68)

The logic we use for filtering / combining message types and choosing how and what to color was inspired by Cline's original code here: https://github.com/cline/cline/blob/main/webview-ui/src/components/chat/task-header/TaskTimeline.tsx#L29 but simplified and updated to work better for both light and dark mode with our tailwind color variables.

Clicking a TimelineMessage will scroll the chat to that message and "blink" it for a moment
![2025-06-18 14 32 19](https://github.com/user-attachments/assets/07b00b05-a0ab-459e-b176-29a9624777dc)

I put the setting to turn on the timeline in a new "Display" settings section:
![image](https://github.com/user-attachments/assets/e18e8b8b-b5af-4e1c-90c9-f66a7db7cabb)

Test plan: Tested manually in a variety of chats / tasks. Tested horizontal scrolling behavior, narrow windows, light/ dark mode

Commit:
chore: Fix Prettier ignore rules

Adds a new `.prettierignore` file to explicitly list directories and files that Prettier should ignore. Updates the `.prettierrc.json` to remove the `ignore` array as it is now handled by the dedicated ignore file. This improves clarity and separation of configuration concerns.

Commit:
feat(storybook): enhance storybook infrastructure with decorators and story reorganization

- Add comprehensive decorator system (withExtensionState, withI18n, withQueryClient, withTheme)
- Create storybook-specific Tailwind CSS configuration
- Reorganize story files from ui/ subdirectory to root stories/
- Add TaskHeader and TaskTimeline stories
- Update preview configuration with new decorators and layout settings
- Add mock data for Cline messages
- Remove obsolete mock utilities
- Update package dependencies for React Query and seedrandom

Commit:
refactor(storybook): complete file reorganization cleanup

- Remove old decorators.tsx file (moved to src/decorators/withTheme.tsx)
- Remove old ui/ subdirectory story files (moved to root stories/ directory)
- Complete the storybook file structure reorganization

Commit:
feat(ui): Add task timeline visualization

Introduces a visual timeline in the task header to display the history of messages within a task. Each message is represented by a colored bar whose width and height are dynamically calculated based on the message's duration and content length.

The timeline allows users to quickly see the flow of a task, identify different message types by color, and click on a bar to scroll the chat view to the corresponding message.

Includes:
- New `TaskTimeline` and `TaskTimelineMessage` components.
- Logic for processing and filtering messages for the timeline.
- Dynamic sizing calculation for timeline bars based on message content and timing.
- Integration into `TaskHeader` and `ChatView`.
- New setting to toggle the visibility of the task timeline.
- Localization strings for the new feature and settings.
- Custom CSS for timeline styling and animations.
- Unit tests for dynamic sizing logic.

Commit:
Clean files

Commit:
Add changeset

Commit:
refactor(storybook): simplify ExtensionState mock with proxy

Replaces the large, manually maintained mock object for ExtensionState in Storybook with a Proxy-based mock creator.

This new approach automatically provides sensible defaults for any property accessed on the mock, significantly reducing maintenance overhead when the ExtensionStateContext type changes.

It also allows for easy overriding of specific properties per story via parameters. Hopefully this should be easy(er) to maintain than the full object since it changes fairly often!

Commit:
PR Cleanup

hassoncs reviewed (commented):
self review

Commit:
refactor(ui): Optimize TaskTimeline rendering with caching

Introduces a `useTimelineCache` hook to memoize the processed timeline data (filtered messages, original indices, and size data).

This prevents redundant recalculations of message processing and size data whenever the `isActive` state changes or the component re-renders for other reasons, improving performance.

Removes the `currentMessageIndex` prop from `TaskTimeline` and `TaskHeader` as the active state is now handled within `TaskTimelineMessage` based on the original index and the overall `isTaskActive` prop.

Adds unit tests for the `useTimelineCache` hook.

Commit:
refactor(ui): Improve TaskTimeline auto-scrolling and layout

Consolidates auto-scroll logic in TaskTimeline to handle both new messages and initial mount.
Removes redundant initial scroll effect.
Adds a small gap between elements in the TaskHeader collapsed state for better spacing.

chrarnoldus commented on webview-ui/src/utils/timeline/timelineMessageProcessing.ts:
```diff
@@ -0,0 +1,75 @@
+ * 3. Filters out unwanted message types
+ * 4. Returns processed messages with original index mapping
+ */
+export function processTimelineMessages(groupedMessages: (ClineMessage | ClineMessage[])[]): {
```
This function seems fairly meaningless

chrarnoldus commented on webview-ui/src/components/chat/TaskTimeline.tsx:
```diff
@@ -0,0 +1,94 @@
+		const timelineMessagesData = useMemo<TaskTimelineMessageData[]>(() => {
+			const { processedMessages, messageToOriginalIndex } = processTimelineMessages(groupedMessages)
+			const currentLength = processedMessages.length
+			const messageSizeData = calculateTaskTimelineSizes(processedMessages)
```
assuming it's O(n log n) or faster and we don't notice any problems I thinks it's fine to not cache it.

chrarnoldus commented:
<img width="468" alt="Screenshot 2025-06-19 at 15 22 03" src="https://github.com/user-attachments/assets/6eb88c07-325e-4378-91ba-fb32295ec73f" />

It looks pretty cool! Maybe a bit cramped once you get a long conversation going?

chrarnoldus committed:
Merge branch 'main' into task-progress-display

chrarnoldus committed:
Fix merge

chrarnoldus committed:
Fix merge

markijbema commented on apps/storybook/src/mockData/clineMessages.ts:
```diff
@@ -0,0 +1,694 @@
+import seedrandom from "seedrandom"
```
is this really necessary? Would regular Math.random give such another view, or hardcoded random numbers?

Commit:
refactor(ui): Improve TaskTimeline performance and caching

Refactors the TaskTimeline component to use a memoized cache for message data and size calculations.

Moves the calculation of raw message data (content length, timing) and size normalization into a new `useTimelineCache` hook. This allows the expensive raw data calculation to be cached incrementally, only recalculating for new messages. The size normalization is still done on the full list but is a much cheaper operation.

Removes the `currentMessageIndex` prop as the active state is now determined within the component based on the last message when the task is active.

Adds a test suite for the new `useTimelineCache` hook.

hassoncs commented on webview-ui/src/utils/timeline/taskTimelineTypeRegistry.ts:
```diff
@@ -0,0 +1,184 @@
+	// Ask types that should be shown (everything except the filtered ones)
+	"ask:tool": {
+		color: taskTimelineColorsByType.SYSTEM_GENERAL_TOOL,
+		translationKey: "tool",
```
I had a similar thought! Will do

hassoncs commented on webview-ui/src/components/chat/TaskTimeline.tsx:
```diff
@@ -0,0 +1,70 @@
+	)
+
+	// Auto-scroll to show the latest message when
+	// new messages are added or on initial mount
```
Yeah, I figured if the llm is busy users wouldn't be actively scrolling back, but maybe that's not actually the case!

Commit:
Address PR comments

Also updated translation keys in `taskTimelineTypeRegistry.ts` to use the `kilocode:` prefix for consistency. Adjusted class name concatenation in `ChatRow.tsx` for improved readability.

Renamed `useTimelineCache` to `useTaskTimelineCache` and `timelineMessageProcessing.ts` to `consolidateMessagesForTimeline.ts` for better clarity and specificity within the task timeline context.

Commit:
Merge branch 'main' into task-progress-display

Commit:
Merge branch 'main' into task-progress-display

Commit:
Fix tests and add missing translations

Commit:
Fix storybook dark styles

Commit:
refactor(timeline): Simplify Caching

Previously, a custom hook `useTaskTimelineCache` was used to perform this processing and caching. Moving this logic into the component simplifies the data flow and reduces the need for a separate hook.

The processing includes:
- Consolidating grouped messages using `consolidateMessagesForTimeline`.
- Calculating message sizes using `calculateTaskTimelineSizes`.
- Determining message colors using `getTaskTimelineMessageColor`.

The `useTimelineCache` hook and its associated test file have been removed. A new test file for `consolidateMessagesForTimeline` has been added.

This change improves code organization and maintainability within the timeline components.
